concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1155 @@
+			givenRetries:       10,
+			wantHTTPStatusCode: http.StatusNoContent,
+			// Close connection before sending so we can check the reconnect behaviour of the NATS connection.
+			wantClosedConnectionBeforeSending: true,
+		},
+	} 
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			testEnv := setupTestEnvironment(t, pkgnats.WithMaxReconnects(tc.givenRetries))
+
+			// subscribe to subject
+			subject := fmt.Sprintf(`"%s"`, testingutils.CloudEventData)
+			// NOTE: we are using the testEnv.natsRecvConnection instead of testEnv.natsSendConnection because the latter will get reconnected based on wantClosedConnectionBeforeSending, this will fail the test when trying to receive a message.
+			done := subscribeToSubject(t, testEnv.natsRecvConnection, subject)
+
+			// create cloudevent with default data (testing.CloudEventData)
+			ce := cloudevents.NewEvent()
+			ce.SetType(testingutils.CloudEventType)
+			err := json.Unmarshal([]byte(testingutils.StructuredCloudEventPayloadWithCleanEventType), &ce)
+			assert.Nil(t, err)
+
+			if tc.wantClosedConnectionBeforeSending {
+				// close connection
+				testEnv.natsSendConnection.Connection.Close()
+				// ensure connection is closed
+				// this is important because we want to test that the connection gets re-established as soon as we send an event
+				assert.True(t, testEnv.natsSendConnection.Connection.IsClosed())
+			}
+
+			// send the event to NATS and asserts that the expectedStatus is returned from NATS
+			status, err := testEnv.natsMessageSender.Send(testEnv.context, &ce)
+			assert.Nil(t, err)
+			assert.Equal(t, tc.wantHTTPStatusCode, status)
+
+			// wait for subscriber to receive the messages
+			err = testingutils.WaitForChannelOrTimeout(done, time.Second*3)
+			assert.NoError(t, err, "Subscriber did not receive the message")
+		})
+	}
+}
+```
+</details>
+
+#### See Also
+- [Table-driven test basics](https://go.dev/blog/subtests)
+
+### Provide test documentation on package level
+
+It is very useful to have some documentation in the test file which describes how the test works from a birds eye view.
+
+Our recommendation is to put this documentation directly at the beginning of the file, before the `package` declaration and before the `imports`.
+The following is an example of the file `components/event-publisher-proxy/pkg/sender/nats_test.go`:
+
+```go
+// Tests in this file are integration tests.
+// They do use a real NATS server using github.com/nats-io/nats-server/v2/server.
+// Messages are sent using NatsMessageSender interface.
+package sender
+```
+
+By following this approach, no godoc will be generated for the package `sender` because it is a test file (ending with `_test` prefix).
+
+### Consistency
+
+#### Assertion Library
+
+The following example is from the event-publisher-proxy. The tests mostly use the `testify/assert` pkg for writing test assertions, but sometimes `testing.T` is used as well.
```

Comment:
```suggestion
The following example is from the event-publisher-proxy. The tests mostly use the `testify/assert` package for writing test assertions, but sometimes `testing.T` is used as well.
```

Reply:
The change found its way into the document. Thank you :)